New draft of Markdown Guide (#202)
withastro/docs: src/pages/en/guides/markdown-content.md

---
layout: ~/layouts/MainLayout.astro
title: Markdown
description: Building Pages in Markdown
---

> Markdown content is commonly used for text-heavy pages like blog posts and documentation. Astro has built-in support for writing in Markdown.

Reviewer: ```suggestion
Markdown is a light weight formatting language that allows anyone to write documents and have them turned into html. Astro takes its built in support for Markdown further, enhancing the developer experience by letting you do more to your Markdown files. 
```

Author: I'll play with this. I personally don't think we have to explain what Markdown content *is*, but I do like emphasizing that you can do more with Markdown in Astro than just Markdown a la mdx.

## Markdown Pages

> Astro treats any `.md` file inside of the `/src/pages` directory as a page. Placing a file in this directory, or any sub-directory, will automatically build a page route using the pathname of the file.

Reviewer: ```suggestion
Astro treats any Markdown (`.md`) file inside of the `/src/pages` directory as its own page. By placing the `.md` file into the `pages/` directory, or any sub-directory, will automatically build the Markdown page as its own route.
```

Author: If you're gonna come at me with a red pen, you best not miss the grammar.  ;)

### Layouts

> Markdown pages have a special front matter property for `layout` that defines the relative path to an `.astro` [layout component](/en/core-concepts/layouts). This component will wrap your Markdown content, providing a page shell and any other page template elements. 

Reviewer: ```suggestion
Markdown pages have a special frontmatter property for `layout` that defines the relative path to an `.astro` [layout component](/en/core-concepts/layouts). This component will wrap around your Markdown content, providing a shell for your content and any other elements displayed on the page. 
```

Author: The nuance is different here. The layout doesn't also wrap the other elements on the page (it contains and uses them to wrap the Markdown content), so this sentence construction can't work as is. I'm inclined to disagree with this! (other than frontmatter)

> 🛎️ Markdown pages are children for layout components. The Markdown content is placed into the layout component's `<slot />` element, and is rendered as HTML.

> All other front matter properties defined in your `.md` page can be passed to the layout component as properties of the `content` object prop.

Reviewer: ```suggestion
All other frontmatter properties defined in your `.md` page will be passed into the layout as properties of the `content` object prop.
```

Author: Interesting distinction. Are the properties "passed" if you don't specify and ask for them, or are they merely "available to be passed"? If a tree falls in a forest...

Reviewer 2: I would accept Fuzzy's distinction: stuff is passed in to a layout from the outside, an outside that doesn't know what the layout wants. It's like I pass you a cup of tea, but you can just drop it on the floor if you don't need it. Still passed it though. 😅

Author: If we go with that, are we confident it should be passed "into" (vs just passed "to"). My gut still says "to" but I can't be sure that's just the language I want to hear, vs correct technical jargon.

(With slot, I think being passed "into" because it does actually GO IN somewhere, by comparison.)

Reviewer 2: I feel like either would be fine but would personally probably also lean towards “to” (just ever so slightly).

Author: OK, I'm taking Fuzzy's WILL and Chris' TO. Go team!

## Markdown Pages

Astro treats any `.md` file inside of the `/src/pages` directory as a page. Placing a file in this directory, or any sub-directory, will automatically build a page route using the pathname of the file. 

📚 Read more about Astro's [file-based routing](/en/core-concepts/routing).
### Basic Example

The easiest way to start using Markdown in Astro is to create a `src/pages/index.md` homepage route in your project. Copy the basic template below into your project, and then view the rendered HTML at the homepage route of your project. Usually, this is at [http://localhost:3000](http://localhost:3000/).

```markdown
---
# Example: src/pages/index.md
title: Hello, World
---
# Hi there!

This is your first markdown page. It probably isn't styled much, although
Markdown does support **bold** and *italics.* 

To learn more about adding a layout to your page, read the next section on **Markdown Layouts.**
```

### Markdown Layouts

> Markdown pages have a special frontmatter property for `layout` that defines the relative path to an `.astro` [layout component](/en/core-concepts/layouts). This component will wrap your Markdown content, providing a page shell and any other included page template elements. 

Reviewer 2: So, is it “an `.astro` component” or “a `.astro` component”? I read “dot-astro” so would assume “a dot-astro component” but this is interesting!

Author: Oh, I've always "said" "an astro component."  I also say "an md file" in my head, too.

Ok, this is one to figure out, as it'll matter site-wide.

Author: Since it's only once in this page, and it's got a link right there to layout components (thereby also indicating official Astro-y), I'm going to just change it to "Astro layout component" !!

```markdown
---
layout: ../layouts/BaseLayout.astro
---
```

> For a layout to work with a Markdown page:
> 1. The page's frontmatter data must be defined as the `content` prop.
> 2. The layout must have a default [`<slot />`](/en/guides/slots) for the Markdown content.

Reviewer 2: I just did [a little test](https://stackblitz.com/edit/github-qmutuh?file=src/layouts/main.astro) and a layout will actually “work” without using `content` and without a `<slot/>`. Obviously it won’t render the Markdown, but it’s not an error.

Given that, I might suggest slightly different wording:

```suggestion
A layout receives data from the Markdown page in two ways:
1. A page's frontmatter data is available as the `content` prop.
2. The Markdown content will be rendered to the layout’s default [`<slot />`](/en/guides/slots).
```

I can imagine some scenarios where you might need _just_ `content` or _just_ a `<slot/>`, so this phrasing doesn’t scare you off doing that.

Author: We're occassionally torn between being *honest* and providing effective guard rails. I think your wording does both, but it feels like it's focusing more on "how this works" vs being something people should generally be sure to do when creating your first layout for a Markdown page. 

I do like a little bit of an implied threat of it not working (as a treat), since I think Fred's original intention here was: "This is what you gotta do to make sure your Layout will render a Markdown page." I actually softened the original wording from "a few rules to follow...."

So, I personally like the more imperative tone from before, but if I strive for better accuracy, and stick to a "rules to follow" take:

A typical layout for Markdown pages:
1. Defines the `content` prop to access the Markdown page's frontmatter data.
2. Includes a default [`<slot />`](/en/guides/slots) to indicate where the page's Markdown content should be rendered.

(I was considering saying one or both are optional, but I think that's the kind of thing people who are really curious, or who have very specific needs, can experiment with on their own. And we can keep this as a good practice for most to keep in mind when building typical pages.)

Thoughts @delucis ?

Reviewer 2: This makes a lot of sense! The one thing I'd change would be to say

1. _Uses_ the `content` prop…

As opposed to “defines”. Technically I guess you are declaring or defining that variable but I think it could be confusing because you “define” the actual data in the page not the layout.

Author: Ooh, I can go one better!

A typical layout for Markdown pages includes:

1. the content prop to access the Markdown page's frontmatter data.
2. a default [<slot />](https://github.com/en/guides/slots) to indicate where the page's Markdown content should be rendered.

??

Reviewer 2: I prefer the one with verbs!

Author: OK, then what about "Declares" instead of defines? I think I really do want to convey that you gotta "const" it there. How does it look/feel in situ?

A typical layout for Markdown pages:

1. Declares the content prop to access the Markdown page's frontmatter data.
2. Includes a default [`<slot />`](https://github.com/en/guides/slots) to indicate where the page's Markdown content should be rendered.

Author: OK, a fuzzy bear said "Uses" too... so I'm outvoted! I give in!

Reviewer 2: Accesses? Although I still prefer “uses”. I was assuming the code example is what people actually look to for usage, so “OK, so I need to _use_ this prop. How? Aha, here’s the example, that is how.”

Reviewer: After taking a small review, to break the impasse, I think the best one that has been proposed, would be Sarah's earlier proposal, this would be adequate of an explanation, personally I feel the less directive this part is, as it is informing the reader of a behavior of astro, it serves little but to add to their understanding. The directive tonality is best served in other parts of the page, 
https://github.com/withastro/docs/pull/202#discussion_r825113900

Author: Gotta go with my Bear here, to break the tie!  ;)

Any time you can common factor out, you probably have a winner.

```astro
---
// src/layouts/BaseLayout.astro

> // 1. The content prop is defined and gives access to frontmatter data

Reviewer 2: Not sure “is defined” is doing much here. More concise would be:

```suggestion
// 1. The content prop gives access to frontmatter data
```

const { content } = Astro.props;
---
<html>
  <head>
  <!-- Add other Head elements here, like styles and meta tags. -->
    <title>{content.title}</title>
  </head>
  <body> 

>   <!-- Add other UI components here, like common headers and footers. -->

Reviewer 2: Indentation nit pick.

```suggestion
    <!-- Add other UI components here, like common headers and footers. -->
```

    <h1>{content.title} by {content.author}</h1>

>     <!-- 2. Rendered HTML passed into the default slot. -->

Reviewer 2: ```suggestion
    <!-- 2. Rendered HTML will be passed into the default slot. -->
```

  </body>
</html>
```

> ### The content prop

Reviewer: ```suggestion
### Props
```

Author: Hmm... normally I like shorter! But, I do kind of like reinforcing that it's "the content prop".

Other opinions??

Reviewer 2: I also feel “Props” would be a sightly different topic. This section is specifically about `content` right?

Author: Yeah, so let's go with The Content Prop for now.

(This is all about getting it ready and making the best choices we can until Fred reads it anyway, to make his life easier! He'll have final say!)

> For Markdown files, the `content` prop also has an `astro` property which holds special metadata about the page such as the complete Markdown `source` and a `headers` object. 

Reviewer: ```suggestion

The `content` prop, contains all your frontmatter properties that are defined onto the page. It also holds special prop named `astro` which contains additional metadata about the page such as the complete Markdown `source` and a `headers` object. 
```

Author: I do like starting the phrase with "The content prop.." so I'll amend.

}
```

> 💡 `astro` and `url` are the only guaranteed properties provided by Astro in the `content` prop. The rest of the object is defined by your front matter variables.

### Variables 

Front matter variables can be used directly in your Markdown as properties of the `frontmatter` object.

```markdown
---
layout: ../layouts/BaseLayout.astro
author: Leon
age: 42
---
# About the Author

> {frontmatter.author} is {frontmatter.age} and lives in Toronto, Canada.

Reviewer: Im pretty certain that this is like:
```suggestion
{author} is {age} and lives in Toronto, Canada.
```

Reviewer: and if we are going with fictional people, could we bring back Andy Kaufman, and have him living on the moon?

Author: I know I snagged this example from somewhere, and this was how it was written. Can't remember if I tested it for myself yet, so I'll do that!

Reviewer 2: I know the blog starter uses `{frontmatter.prop}` and that works.

Author: That must be where I grabbed it from, and I just double checked, and that's in fact what they do.

(I also kind of hope it works that way, because it's a nice symmetry with the `content.` object stuff.)

```

### Components

> You can import components into your Markdown file with `setup` and use them along with Markdown syntax. The `frontmatter` object is also available to any imported components.

Reviewer: ```suggestion
You can import components into your Markdown file with `setup` and use them alongside your Markdown content. The `frontmatter` object is also available to any imported components.
```

Author: I'll buy this. I was really thinking of it from a syntax/writing/mix-and-match standpoint, but I have no issues with content. With Markdown, the content *is* closely tied to the syntax.

```markdown
---
layout: ../layouts/BaseLayout.astro
setup: | 
  import Author from '../../components/Author.astro'
  import Biography from '../components/Biography.jsx'
author: Leon
---
# About the Author

<Author name={frontmatter.author}/>

> <Biography client:visible >

Reviewer 2: Stray space here.

```suggestion
<Biography client:visible>
```

The rest of this looks good to me!

{frontmatter.author} lives in Toronto, Canada and enjoys photography.
</Biography>

```

### Draft Pages

> `draft: true` is an optional front matter value that will mark an individual `.md` page or post as "unpublished." By default, this page will be excluded from the site build.

Reviewer: ```suggestion
`draft: true` is an optional frontmatter value that will mark an individual `.md` page or post as "unpublished." This will result in the page being excluded from the final build of the site.
```

Author: The reason this says "by default" is because there *is* a configuration setting wherein you can specify that you want draft posts built anyway. But the default is to not build them, and you have to opt-in to build them.  (frontmatter is indeed, one word, though! I'll do a full document check for that!)

> Markdown pages with `draft: false` or without the `draft` property are unaffected and will be built.

Reviewer: ```suggestion
Markdown pages without the `draft` property or those with `draft: false` are unaffected and will be included into the final build.
```

Author: Again, let's compromise. I do like putting "without the draft property" first. Sounds better. But I'm gonna pull grammar rank and say it should be "included in the final build"
